hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/TestLatencyTracker.java

```diff
@@ -0,0 +1,479 @@
+    AbfsHttpOperation httpOperation = new AbfsHttpOperation(url, "GET", new ArrayList<AbfsHttpHeader>());
+
+    for (int i=0; i < numTasks; i++) {
+      Callable<Integer> c = new Callable<Integer>() {
+        @Override
+        public Integer call() throws Exception {
+          // test latency tracking when aggregate latency numbers are also passed
+          latencyTracker.recordClientLatency(Instant.now(), "oneOperationCaller", "oneOperationCallee", true, Instant.now(), TEST_AGGREGATE_LATENCY, httpOperation);
+          return 0;
+        }
+      };
+      tasks.add(c);
+    }
+
+    for (Future<Integer> fr: executorService.invokeAll(tasks)) {
+      fr.get();
+    }
+
+    for (int i=0; i < numTasks; i++) {
+      latencyDetails = latencyTracker.getClientLatency();
+      Assert.assertNotNull("LatencyTracker should return non-null record", latencyDetails);
+      Assert.assertTrue("Latency record should be in the correct format", Pattern.matches(
+              "h=[^ ]* t=[^ ]* a=bogusFilesystemName c=bogusAccountName cr=oneOperationCaller ce=oneOperationCallee r=Succeeded l=[0-9]+"
+                      + " ls=[0-9]+ lc=" + TEST_AGGREGATE_LATENCY + " s=0 e= ci=[^ ]* ri=[^ ]* bs=0 br=0 m=GET u=http%3A%2F%2Fwww.microsoft.com%2FbogusFile", latencyDetails));
+    }
+  }
+
+  @Test
+  public void verifyRecordingSingletonLatencyIsCheapWhenDisabled() throws Exception {
+    // when latency tracker is disabled, we expect it to take time equivalent to checking a boolean value
+    final long maxLatencyWhenDisabledMs = 1;
+    final long minLatencyWhenDisabledMs = 0;
+    final long numTasks = 1000;
+    long aggregateLatency = 0;
+    LatencyTracker latencyTracker = new LatencyTracker(accountName, filesystemName, false);
+
+    ExecutorService executorService = Executors.newCachedThreadPool();
+    List<Callable<Long>> tasks = new ArrayList<Callable<Long>>();
+    final AbfsHttpOperation httpOperation = new AbfsHttpOperation(url, "GET", new ArrayList<AbfsHttpHeader>());
+
+    for (int i=0; i < numTasks; i++) {
+      Callable<Long> c = new Callable<Long>() {
+        @Override
+        public Long call() throws Exception {
+          Instant startRecord = Instant.now();
+
+          try{
+          } finally {
+            latencyTracker.recordClientLatency(startRecord, "oneOperationCaller", "oneOperationCallee", true, httpOperation);
+          }
+
+          long latencyRecord = Duration.between(startRecord, Instant.now()).toMillis();
+          LOG.debug("Spent {} ms in recording latency.", latencyRecord);
+          return latencyRecord;
+        }
+      };
+      tasks.add(c);
+    }
+
+    for (Future<Long> fr: executorService.invokeAll(tasks)) {
+      aggregateLatency += fr.get();
+    }
+
+    double averageRecordLatency = aggregateLatency/numTasks;
+    Assert.assertTrue(String.format("Average time for recording singleton latencies, %s ms should be in the range [%s, %s).",
```

Comment:
we're moving to AssertJ assertions where appropriate; this looks like somewhere where could try an Assertions.assertThat() ... its very informative on failures

Reply:
Added AssertJ to the hadoop-azure module and used it in the tests here.